File: backend/app/models/pessoa.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class PessoaCreate(BaseModel):
    nome: str = Field(..., min_length=2, max_length=150)
    natural_de: Optional[str] = Field(None, max_length=100)
    provincia: Optional[str] = Field(None, max_length=100)
    estado_civil: Optional[str] = Field(None, max_length=50)
    profissao: Optional[str] = Field(None, max_length=100)
    residencia: Optional[str] = Field(None, max_length=200)
    comunidade: Optional[str] = Field(None, max_length=150, description="Comunidade/paróquia a que pertence, se não for a nossa")
    contacto: Optional[str] = Field(None, max_length=50)


class PessoaUpdate(BaseModel):
    nome: Optional[str] = Field(None, min_length=2, max_length=150)
    natural_de: Optional[str] = Field(None, max_length=100)
    provincia: Optional[str] = Field(None, max_length=100)
    estado_civil: Optional[str] = Field(None, max_length=50)
    profissao: Optional[str] = Field(None, max_length=100)
    residencia: Optional[str] = Field(None, max_length=200)
    comunidade: Optional[str] = Field(None, max_length=150)
    contacto: Optional[str] = Field(None, max_length=50)


class PessoaOut(BaseModel):
    id: str
    nome: str
    natural_de: Optional[str] = None
    provincia: Optional[str] = None
    estado_civil: Optional[str] = None
    profissao: Optional[str] = None
    residencia: Optional[str] = None
    comunidade: Optional[str] = None
    contacto: Optional[str] = None
    criado_em: datetime


class VinculoPessoaOut(BaseModel):
    """Um catequisando que referencia esta pessoa, e em que papel."""
    catequisando_id: str
    catequisando_nome: str
    papel: str  # "pai" | "mae" | "padrinho" | "madrinha"

File: frontend/lib/models/pessoa.dart
class Pessoa {
  final String id;
  final String nome;
  final String? naturalDe;
  final String? provincia;
  final String? estadoCivil;
  final String? profissao;
  final String? residencia;
  final String? comunidade;
  final String? contacto;

  Pessoa({
    required this.id,
    required this.nome,
    this.naturalDe,
    this.provincia,
    this.estadoCivil,
    this.profissao,
    this.residencia,
    this.comunidade,
    this.contacto,
  });

  factory Pessoa.fromJson(Map<String, dynamic> json) => Pessoa(
        id: json['id'] as String,
        nome: json['nome'] as String,
        naturalDe: json['natural_de'] as String?,
        provincia: json['provincia'] as String?,
        estadoCivil: json['estado_civil'] as String?,
        profissao: json['profissao'] as String?,
        residencia: json['residencia'] as String?,
        comunidade: json['comunidade'] as String?,
        contacto: json['contacto'] as String?,
      );
}

/// Um catequisando que referencia uma Pessoa, e em que papel (ex: para
/// mostrar "afilhados" de um padrinho, ou "filhos" de um pai/mãe).
class VinculoPessoa {
  final String catequisandoId;
  final String catequisandoNome;
  final String papel;

  VinculoPessoa({required this.catequisandoId, required this.catequisandoNome, required this.papel});

  factory VinculoPessoa.fromJson(Map<String, dynamic> json) => VinculoPessoa(
        catequisandoId: json['catequisando_id'] as String,
        catequisandoNome: json['catequisando_nome'] as String,
        papel: json['papel'] as String,
      );

  String get papelRotulo {
    switch (papel) {
      case 'pai':
        return 'Filho(a)';
      case 'mae':
        return 'Filho(a)';
      case 'padrinho':
        return 'Afilhado(a)';
      case 'madrinha':
        return 'Afilhado(a)';
      default:
        return papel;
    }
  }
}
